syslog/drainer.go

```diff
@@ -158,12 +166,26 @@ func (d *drainer) formatMessage(line string, tag string) ([]byte, error) {
 	return binary, nil
 }
 
+func (d *drainer) resetRetries() {
+	d.connAttempts = 0
+	d.sleepSeconds = 1
+}
+
+func (d *drainer) incrementAttempts() {
+	d.connAttempts++
+	d.sleepSeconds = d.sleepSeconds << 1
+}
+
 func (d *drainer) ensureConnection() {
 	for d.conn == nil {
+		d.incrementAttempts()
 		conn, err := d.dialFunction()
 		if err != nil {
-			d.errorLogger.Printf("Error connecting: %s \n", err.Error())
-			time.Sleep(time.Second)
+			if d.maxRetries > 0 && d.connAttempts > d.maxRetries {
+				d.errorLogger.Fatalln("Failed to connect to syslog server. Exiting now.")
+			}
+			d.errorLogger.Printf("Error connecting: '%s', Retry: %d, Will retry in %d seconds.\n", err.Error(), d.connAttempts, d.sleepSeconds)
```

Comment:
Labelling connection attempts as "Retry" seems a bit misleading, like if I saw `Retry: 1` I think that I would expect that to be the second attempt.

Perhaps we could rename the label?

While we're in there I'm also not a huge fan of this comma separated logging. What do you think about something more like: `"Error connecting on attempt %d: %s. Will retry in %d seconds.\n"`